### ISSUE TITLE
[Feature Request/Discussion] CollectivePermute API

### ISSUE THREAD

**Issue by DyeKuu** [NVIDIA/nccl#943]
Hello,

I'd like to propose a feature request and start a discussion around introducing a `CollectivePermute` API to NCCL (or a simplified version that does one-step cyclic shift). This distributive primitive is already in use within frameworks like XLA, and exposing this API directly through NCCL could offer several key benefits.

The motivation is to provide users with more fine-grained control, especially in the context of distributed training tensor parallelism, where a blocking AllGather is followed by a matmul operation. While the existing AllGather operation serves its purpose, I believe the potential for overlapping computation in-between operations could be better achieved through a CollectivePermute operation. An illustrative way to think about this is: `AllGather = forloop([CollectivePermute, concat])`. Indeed a recent work did this (https://dl.acm.org/doi/pdf/10.1145/3567955.3567959) that decomposes `AllGather` and `ReduceScatter` using `CollectivePermute` in XLA and overlaps matmul in-between the communication.

Though for now we could mimic this behavior based on `send/recv`. But it's a bit hard for developers building upon `torch.distributed` API where it only binds to existing nccl API.

GSPMD: https://arxiv.org/pdf/2105.04663.pdf
XLA CollectivePermute: https://www.tensorflow.org/xla/operation_semantics#collectivepermute

**Comment by sjeaugey** [comment]
I'm not sure I understand what prevents us from providing a CollectivePermute primitive at the Python API level, implementing it with ncclSend/ncclRecv at the C++ level. Is there some matmul we'd want NCCL to to internally, pipelined with communication?

**Comment by DyeKuu** [comment]
> I'm not sure I understand what prevents us from providing a CollectivePermute primitive at the Python API level, implementing it with ncclSend/ncclRecv at the C++ level

We could surely do this on python side. It's more like potential performance reason if this primitive is pieced together with existing API calls at python level that I'm not sure.

> Is there some matmul we'd want NCCL to to internally, pipelined with communication?

This is the best world that I could hope for. Ideally it would be great to have function like below with high perf:

```python
def collective_permute_all_gather_matmul(input_tensor, weight):
    """
    Performs a simulated AllGather using collective_permute, followed by a matrix multiplication.
    Returns the accumulated matrix multiplication result.
    """
    rank = dist.get_rank()
    world_size = dist.get_world_size()

    # Initialize an output tensor
    output = [torch.zeros_like(input_tensor @ weight) for _ in range(world_size)]

    # Loop for world_size times
    for i in range(world_size):
        # Perform matmul and accumulate
        output[i] = input_tensor @ weight

        # Perform collective permute
        input_tensor, reqs = collective_permute_async(input_tensor)
        for req in reqs:
            req.wait()

    return torch.cat(output)
```

But I'm not sure if this is asking too much as I presume doing matmul inside nccl is out of its scope. So for the first step I would ask about the possibility of having such `collective_permute_async` in nccl.

**Comment by sjeaugey** [comment]
I'm still failing to understand how `collective_permute_async` would be implemented in NCCL and why we need a dedicated API for that specific use case if we can implement it efficiently with ncclSend and ncclRecv.

**Comment by DyeKuu** [comment]
> if we can implement it efficiently with ncclSend and ncclRecv.

I agree. To be honest this is also the path I was exploring:  Using `torch.distributed` to write such primitive. My preliminary trial is to do something like

```python
def collective_permute_async(recv_tensor, tensor, src_rank, dest_rank):
    """
    Performs a cyclic left shift on the tensor across ranks.
    """
    # Create send and receive operations
    # P2POp takes global ranks as arguments
    send_op = dist.P2POp(dist.isend, tensor, dest_rank)
    recv_op = dist.P2POp(dist.irecv, recv_tensor, src_rank)

    # Execute the operations in a batch and return the requests
    reqs = dist.batch_isend_irecv([send_op, recv_op])

    return reqs
```

But I observe some performance degradation when I try to compare the "collective_permute in forloop" with "AllGather" if I don't do any computation inside the for loop. My profiler shows that calling P2P multiple times is quite slower than a single AllGather call. So I'm not really sure whether building upon python API for such distributive primitive is the right level to work on if I care most about the performance. 

Another possible way I can think of is to work on torch's level, to write something like
```cpp
torch::Tensor collective_cyclic_shift(torch::Tensor recv_tensor, torch::Tensor tensor, int src_rank, int dest_rank, ncclComm_t comm, cudaStream_t stream) {
...
}
```
in cpp using existing ncclSend/ncclRecv then bind it to Python. Since torch doesn't expose the NCCL Communicator to python side, I presume that this may require to modify torch's source code.

**Comment by DyeKuu** [comment]
A naïve question. In the send/recv I'm using uni-directional communication only: am I just using half of the bandwidth of NVLink?

**Comment by sjeaugey** [comment]
If each GPU is receiving from the previous one, and sending to the next one, then each NVLink is used in both directions. That's because the same link is used but sending and receiving in an NVSwitch system. However, if you were on a system with direct connections between GPUs (direct NVLink connections for example) then you would indeed need multiple rings to maximize the bandwidth.
And in that case, I could see how NCCL has more information regarding which rings to use but I can't think of a way to present an API that would let you do what you want.

**Comment by stephenmsachs** [comment]
### Issue Cleanup: Helping Us Focus on Current Challenges

  We're [reviewing](https://github.com/NVIDIA/nccl/discussions/1761) older issues to ensure we prioritize the most relevant and active ones. Since this issue hasn't seen updates in over 6 months, we'll be closing it for now.

  ,*This change helps us focus our efforts on addressing any current issues our users are facing.* If this issue still affects you, please don't hesitate to reopen it with a quick update (e.g., "Still relevant on [version=X]").
  Thanks for your understanding and for contributing to NCCL.
